### PR TITLE
Introduce Generator in RandomVariable

### PR DESCRIPTION
**Pull request by kc611**
This PR will serve as the entry point of Numpy's new `Generator` API for use in `RandomVariable` class.

### CODE REVIEW

File: aesara/tensor/random/type.py [hunk 1]
```diff
@@ -31,31 +45,46 @@ def filter(cls, data, strict=False, allow_downcast=None):
 
     @staticmethod
     def is_valid_value(a, strict):
-        if isinstance(a, np.random.RandomState):
+        if isinstance(a, np.random.Generator) or isinstance(a, np.random.RandomState):
```
Comment (brandonwillard) [inline aesara/tensor/random/type.py:48]:
```suggestion
        if isinstance(a, (np.random.Generator, np.random.RandomState)):
```

**Review by brandonwillard — COMMENTED** [review]
Looks good!

Before we merge this, we change all internal uses of `RandomState`&mdash;that aren't used to explicitly test `RandomState`&mdash;to the new `Generator`, just so that the library is consistent about using the new API.

**Comment by kc611** [comment]
The convention I'm following for these tests migrations is:

`RandomState` -> `Generator` intitalization.

`.rand` -> `.random` function

`.randint` -> `.integers` function

`.random_sample` -> `.random` function

`randn` -> `.standard_normal` function

**Comment by brandonwillard** [comment]
> The convention I'm following for these tests migrations is:
> 
> `RandomState` -> `Generator` intitalization.
> 
> `.rand` -> `.random` function
> 
> `.randint` -> `.integers` function
> 
> `.random_sample` -> `.random` function
> 
> `randn` -> `.standard_normal` function

Sounds correct to me.

File: aesara/link/jax/dispatch.py [hunk 2]
```diff
@@ -108,6 +108,13 @@ def jax_typify_RandomState(state, **kwargs):
     return state
 
 
+@jax_typify.register(Generator)
+def jax_typify_Generator(state, **kwargs):
+    state = state.__getstate__()
+    state["bit_generator"] = numpy_bit_gens[state["bit_generator"]]
```
Comment (kc611) [inline aesara/link/jax/dispatch.py:114]: Regarding the Jax conversion. I was thinking maybe we could generate a `PRNG` over here (and do the same with `RandomState`) and store it in the `state` dictionary as `jax_key`.

And in the `RandomVariable` conversion down below we could update this `jax_key` instead of the Numpy's state key.

Comment (brandonwillard) [reply aesara/link/jax/dispatch.py:114]: As long as `jax_key` stays localized to the JAX compilation process (i.e. we don't have to alter the generic `RandomState`/`Generator` types to explicitly account for it), we should be fine.

**Comment by kc611** [comment]
In `tests.unittest_tools` the function `set_seed()` :

https://github.com/pymc-devs/aesara/blob/fbde872ef7cb6098b23ad442b82613eb41702de3/tests/unittest_tools.py#L54

Depends upon the Numpy's `random.seed()` function, which should probably be removed since it internally seeds the `RandomState` only. (And it's use is not recommended by Numpy either). 

Instead we can just define a Generator as `np.random.default_rng(utt.fetch_seed())` where-ever the consistency is required between random arrays. And having a externally accessible Generator and calling random methods from that is much more convenient than internally using `set_seed()` and calling random function using `numpy.random` module.

**Comment by kc611** [comment]
I think the current failing test is caused by some kind of inconsistency between `rng` states. I just can't figure out where. (Maybe it's related to removing the unit test module's `set_seed` method?) And I can't replicate this issue locally. (The tests pass, locally).

@brandonwillard Any guesses?

**Comment by brandonwillard** [comment]
> Instead we can just define a Generator as `np.random.default_rng(utt.fetch_seed())` where-ever the consistency is required between random arrays.

Definitely; however, I would also add that tests should always use their own explicitly seeded `Generator`s.  This is another big problem we've inherited from Theano: we have too many tests that use a global/shared RNG, which makes all the tests inherently coupled.  One can see this clearly when trying to refactor, because the RNG state-sensitive tests start to fail when new tests are added or moved and cause the test order&mdash;and, thus, RNG state&mdash;to change.

**Comment by brandonwillard** [comment]
> I think the current failing test is caused by some kind of inconsistency between `rng` states. I just can't figure out where.

I may have accidentally provided a reason in my previous comment.

File: tests/scan/test_basic.py [hunk 3]
```diff
@@ -806,8 +806,8 @@ def get_sum_of_grad(input0, input1):
         # Call verify_grad to ensure the correctness of the second gradients
         floatX = config.floatX
         inputs_test_values = [
-            np.random.random((2, 3)).astype(floatX),
-            np.random.random(3).astype(floatX),
+            np.random.default_rng(utt.fetch_seed()).random((2, 3)).astype(floatX),
+            np.random.default_rng(utt.fetch_seed()).random(3).astype(floatX),
```
Comment (brandonwillard) [inline tests/scan/test_basic.py:810]: It looks like this will reset the state.  Try creating `rng = np.random.default_rng(utt.fetch_seed())` once and using `rng` for both these cases.

**Review by brandonwillard — CHANGES_REQUESTED** [review]

File: tests/scan/test_basic.py (hunk 3 above)
Comment (kc611) [reply tests/scan/test_basic.py:810]: Already tried, and also tried passing that rng all the way to `verify_grad` it still fails. And strangely I can't replicate any of this locally. I have to commit and 'uncommit' each time.

Comment (brandonwillard) [reply tests/scan/test_basic.py:810]: Well, let's keep that change anyway. 

Otherwise, the failures could simply be due to the underlying RNG change generating a new set of samples, so we can update the test for this difference (e.g. the accepted error ranges and such), find a more advantageous seed, or something else.  Changing the seed is the least preferrable, though.

**Comment by kc611** [comment]
I think the migration to `Generator`s is mostly done. 
 
 A few compromises I needed to make:
 
Had to increase the tolerance over here for gradient verification: (increased from 0.01 to 0.05, tests will also pass if it's set to 0.02, but there may be flaky failures)
 https://github.com/pymc-devs/aesara/blob/c32c1d34f9ea7e11e877bd454cb9b08305812720/tests/unittest_tools.py#L79

Had to increase tolerance in test/tensor/test_blas for GEMM(increased form 1e-8 to 2e-8, but this one is seed based test and I'm pretty sure we can find a seed which will still have a tolerance of 1e-08)
https://github.com/pymc-devs/aesara/blob/c32c1d34f9ea7e11e877bd454cb9b08305812720/tests/tensor/test_blas.py#L674

The Jax tests for `Generators` are marked as `xfail` for now, it'll be better if we do that in a separate PR. 

Other than this the commits need to be organized (not sure how I should do that, maybe put all the work in a single commit? ). And regarding code coverage I think most of the negative coverage (about `-0.06`) comes from the dictionaries defined in `tensor/type.py`. Other than the ones added in `tests/tensor/test_type` I don't think there's need for extra tests.

**Comment by brandonwillard** [comment]
I just rebased and updated the NumPy requirement (so that these changes won't fail for older versions).

**Review by brandonwillard — COMMENTED** [review]
After disabling the `xfail` on `tests.link.test_jax.test_random_generators`, I noticed that it fails due to an inability to convert an extremely long `int` to NumPy's `int64`.  

From what I can tell, this large `int` is coming from the `Generator` state `dict` (e.g. either the `state` or `inc` entry under the `state` key).

We need to check the handling of these `dict` representations and make sure they can be used inside a compiled Aesara function.

**Comment by kc611** [comment]
> We need to check the handling of these dict representations and make sure they can be used inside a compiled Aesara function.

Is the current approach (splitting that integer into individual digits) a viable one ?

**Review by brandonwillard — CHANGES_REQUESTED** [review]
I've rebased and updated the JAX random variable tests so that they check the sample means; however, they won't pass, because there appears to be a dtype issue somewhere that's causing&mdash;for instance&mdash;normal random variables to take an integer dtype.

**Comment by brandonwillard** [comment]
> Is the current approach (splitting the big integer into individual digits) a viable one ?

I've updated the approach a little bit, and it should work well enough for now.

The changes involve splitting and recombining `Generator`'s 128 bit `int`s in terms of 32 bit `int`s.  Since JAX's RNG states appear to consist of two 32 bit `int`s, this approach seems compatible.

**Review by brandonwillard — APPROVED** [review]

**Review by brandonwillard — CHANGES_REQUESTED** [review]
We need to update the `RandIntRV` `Op`, because `Generator.randint` doesn't work.

We could rename it to `IntegersRV` and make it use `RandomState.randint` or `Generator.integers`&mdash;depending on the undering `rng`&mdash;however, this doesn't look immediately possible with the current `RandomGeneratorType`.

I think the easiest way to solve this is to have distinct Aesara `Type`s for each RNG type, because then we can simply use the `Type` within `Op` logic (and other places).

At the very least, both `Type`s could share an abstract base class that is effectively the current `RandomGeneratorType`.

**Comment by kc611** [comment]
`RandIntRV` currently uses `.integers`, Sure we can rename that to `IntegersRV`. But I don't see any other use case for two different RNG `Type`s outside of this. (Since the interfaces are so similar except for `.random`/`.integers` method).

One other thing we can do is add a case for `IntegersRV` in `RandomVariable` which will just change the `name` to `.random` if underling RNG is a `RandomState` which can be done either during initialization or the actual sampling. Since Aesara as well as Numpy is now `Generators` based and keeping `RandomStates` for backwards compatibility, I think having different `Types` would be overkill if this is the only use case. Or is it ?

**Comment by brandonwillard** [comment]
> But I don't see any other use case for two different RNG `Type`s outside of this. (Since the interfaces are so similar except for `.random`/`.integers` method).

This will likely reappear when we/if we get Numba support for `Generator`, since then we'll need to be able to determine that the underlying RNG is truly a `Generator`.

Currently, we're hiding the type of the underlying non-symbolic object with our dual-purpose symbolic `Type`, and that invites all sorts of trouble (and effectively breaks the implicit API/object contract in certain ways).

**Comment by brandonwillard** [comment]
There's a very weird [segmentation fault](https://github.com/aesara-devs/aesara/pull/420/checks?check_run_id=2907426448#step:6:44) showing up in the Numba tests.  I don't see how this is related to these `Generator` changes, and I can't reproduce it locally.

I added a test commit (to be removed) that only runs those tests, and it appears to be very reproducible in the CI environment.

**Comment by brandonwillard** [comment]
The segmentation fault is fixed in https://github.com/aesara-devs/aesara/pull/498.
